### PR TITLE
[#1475] Adopt transaction data requests

### LINKED ISSUE #1475: Adopt transaction data requests

### PR DESCRIPTION
**Pull request by LukasKorba**
Closes #1475 

- RustBackend extended for the new 2 rust methods 'transactionDataRequests' and 'setTransactionStatus'
- lightwalletservice extended to add a new method 'getTaddressTxids'
- Enhance action has been refactored to handle transactionDataRequests
- Error codes for specific rust and service errors defined
- Checkpoints added
- Code cleanup

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Block/Enhance/BlockEnhancer.swift [hunk 1]
```diff
@@ -60,21 +60,24 @@ struct BlockEnhancerImpl {
...
-        logger.debug("Decrypting and storing transaction id: \(transactionID) block: \(block)")
-
-        try await rustBackend.decryptAndStoreTransaction(
-            txBytes: fetchedTransaction.raw.bytes,
-            minedHeight: Int32(fetchedTransaction.minedHeight)
-        )
+        if fetchedTransaction.minedHeight == -1 {
+            try await rustBackend.setTransactionStatus(txId: fetchedTransaction.rawID, status: .txidNotRecognized)
+        } else if fetchedTransaction.minedHeight == 0 {
+            try await rustBackend.setTransactionStatus(txId: fetchedTransaction.rawID, status: .notInMainChain)
```
Comment (nuttycom) [inline Sources/ZcashLightClientKit/Block/Enhance/BlockEnhancer.swift:74]: Whether to call `setTransactionStatus` or `decryptAndStoreTransaction` should depend upon the type of the request received, not on the mined height.

Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Block/Enhance/BlockEnhancer.swift]: We completely updated the code together on a pairing.

File: Sources/ZcashLightClientKit/Model/TransactionDataRequest.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//
+//  Created by Lukáš Korba on 08-15-2024.
+//
+
+import Foundation
+
+struct SpendsFromAddress: Equatable {
+    let address: String
+    let blockRangeStart: UInt32
+    let blockRangeEnd: Int64
```
Comment (nuttycom) [inline Sources/ZcashLightClientKit/Model/TransactionDataRequest.swift:13]: `blockRangeEnd` shoudl be `UInt32?`. After the `BlockHeight` refactoring, `blockRangeStart: BlockHeight` and `blockRangeEnd: BlockHeight?`

**Review by nuttycom — CHANGES_REQUESTED** [review]
The only problem I see is the `Int64` being used for the address request end height.

File: Sources/ZcashLightClientKit/Entity/TransactionEntity.swift [hunk 3]
```diff
@@ -94,7 +94,7 @@ public enum ZcashTransaction {
     /// Used when fetching blocks from the lightwalletd
     struct Fetched {
         public let rawID: Data
-        public let minedHeight: BlockHeight
+        public let minedHeight: UInt32?
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Entity/TransactionEntity.swift:97]: Agreed, let's make it in the next iteration. https://github.com/Electric-Coin-Company/zcash-swift-wallet-sdk/issues/1477

File: Sources/ZcashLightClientKit/Model/TransactionDataRequest.swift (hunk 2 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Model/TransactionDataRequest.swift]: Set to UInt32? until #1477 is implemented.

**Review by nuttycom — APPROVED** [review]
utACK ad30b247b091f75bc5dc8bab4755f3cecb25395d

**Pull request merged**